File: Tutorials/Codes/System_Dynamics_Course/Chapter20/Lesson4/Chapter20_Lesson4.m
% Chapter20_Lesson4.m
%{
Chapter 20 — Chaos, Complex Dynamics, and Computational Tools
Lesson 4 — Computational Tools

MATLAB demonstration:
1) Lorenz ODE with ode45 (nonstiff) and ode15s (stiff-capable)
2) Event detection for Poincaré section (x=0 crossings, increasing direction)
3) Logistic map bifurcation tail data + Lyapunov exponent curve
4) Optional: Programmatic creation of a simple Simulink model for Lorenz

Outputs: CSV files written to a local folder "outputs_ch20_l4_matlab".
%}

function Chapter20_Lesson4()
  outDir = "outputs_ch20_l4_matlab";
  if ~exist(outDir, 'dir'), mkdir(outDir); end

  % Parameters
  p.sigma = 10.0;
  p.rho   = 28.0;
  p.beta  = 8.0/3.0;

  x0 = [1;1;1];
  tspan = [0, 40];

  % --- ODE45 with event detection ---
  opts = odeset('RelTol',1e-9,'AbsTol',1e-12,'MaxStep',0.02,'Events',@(t,x) ev_x0(t,x,p));
  [t, x, te, xe] = ode45(@(t,x) lorenz_rhs(t,x,p), tspan, x0, opts);

  writematrix([t, x], fullfile(outDir, "lorenz_traj.csv"));

  if ~isempty(xe)
    writematrix(xe, fullfile(outDir, "lorenz_poincare_x0.csv"));
  end

  % --- Compare with ode15s (useful when stiffness appears) ---
  % Lorenz is not stiff in this parameter set, but we show usage:
  opts15 = odeset('RelTol',1e-9,'AbsTol',1e-12,'MaxStep',0.02);
  [t15, x15] = ode15s(@(t,x) lorenz_rhs(t,x,p), tspan, x0, opts15); %#ok<ASGLU>
  % (Not exported to keep files minimal.)

  % --- Logistic map bifurcation data ---
  rVals = linspace(2.5, 4.0, 400);
  nTrans = 800; nKeep = 120;
  bif = zeros(numel(rVals)*nKeep, 2);
  idx = 1;
  for i = 1:numel(rVals)
    r = rVals(i);
    x = 0.2;
    for k = 1:nTrans, x = r*x*(1-x); end
    for k = 1:nKeep
      x = r*x*(1-x);
      bif(idx,:) = [r, x];
      idx = idx + 1;
    end
  end
  writematrix(bif, fullfile(outDir, "logistic_bifurcation.csv"));

  % --- Logistic map Lyapunov curve ---
  rVals2 = linspace(2.8, 4.0, 200);
  ly = zeros(numel(rVals2), 2);
  for i = 1:numel(rVals2)
    r = rVals2(i);
    ly(i,:) = [r, logistic_lyapunov(r, 0.2, 1000, 4000)];
  end
  writematrix(ly, fullfile(outDir, "logistic_lyapunov.csv"));

  % --- Optional: build a Simulink model programmatically ---
  % If you prefer manual building, create 3 Integrator blocks and one MATLAB Function
  % that outputs dx,dy,dz and wire them as xdot -> Integrator -> x.
  if exist('simulink', 'file')
    try
      build_simulink_lorenz(p);
    catch ME
      fprintf("Simulink model build failed: %s\n", ME.message);
    end
  end

  disp("Done. Outputs written to: " + outDir);
end

function dx = lorenz_rhs(~, x, p)
  X = x(1); Y = x(2); Z = x(3);
  dx = [ p.sigma*(Y - X);
         X*(p.rho - Z) - Y;
         X*Y - p.beta*Z ];
end

function [value, isterminal, direction] = ev_x0(~, x, ~)
  value = x(1);        % x == 0
  isterminal = 0;      % do not stop
  direction = +1;      % negative -> positive only
end

function lam = logistic_lyapunov(r, x0, nTrans, n)
  x = x0;
  for k = 1:nTrans, x = r*x*(1-x); end
  s = 0.0;
  for k = 1:n
    x = r*x*(1-x);
    d = abs(r*(1 - 2*x));
    if d < 1e-300, d = 1e-300; end
    s = s + log(d);
  end
  lam = s / n;
end

function build_simulink_lorenz(p)
  mdl = "Chapter20_Lesson4_Lorenz";
  if bdIsLoaded(mdl), close_system(mdl, 0); end
  new_system(mdl); open_system(mdl);

  % Add 3 integrators for x,y,z
  add_block('simulink/Continuous/Integrator', mdl + "/IntX", 'Position',[200 60 230 90]);
  add_block('simulink/Continuous/Integrator', mdl + "/IntY", 'Position',[200 130 230 160]);
  add_block('simulink/Continuous/Integrator', mdl + "/IntZ", 'Position',[200 200 230 230]);

  % Set initial conditions
  set_param(mdl + "/IntX", 'InitialCondition', '1');
  set_param(mdl + "/IntY", 'InitialCondition', '1');
  set_param(mdl + "/IntZ", 'InitialCondition', '1');

  % MATLAB Function block for derivatives
  add_block('simulink/User-Defined Functions/MATLAB Function', mdl + "/LorenzF", 'Position',[60 110 140 180]);
  code = sprintf([ ...
    "function dx = f(x)\n" ...
    "%% x = [X;Y;Z]\n" ...
    "sigma = %.17g; rho = %.17g; beta = %.17g;\n" ...
    "X = x(1); Y = x(2); Z = x(3);\n" ...
    "dx = [ sigma*(Y-X); X*(rho-Z)-Y; X*Y - beta*Z ];\n" ...
    "end\n"], p.sigma, p.rho, p.beta);
  set_param(mdl + "/LorenzF", 'Script', code);

  % Mux and Demux
  add_block('simulink/Signal Routing/Mux', mdl + "/Mux", 'Inputs','3','Position',[160 110 180 180]);
  add_block('simulink/Signal Routing/Demux', mdl + "/Demux", 'Outputs','3','Position',[150 60 170 230]);

  % Connect integrator outputs -> mux -> function -> demux -> integrator inputs
  add_line(mdl, "IntX/1", "Mux/1");
  add_line(mdl, "IntY/1", "Mux/2");
  add_line(mdl, "IntZ/1", "Mux/3");
  add_line(mdl, "Mux/1", "LorenzF/1");
  add_line(mdl, "LorenzF/1", "Demux/1");
  add_line(mdl, "Demux/1", "IntX/1");
  add_line(mdl, "Demux/2", "IntY/1");
  add_line(mdl, "Demux/3", "IntZ/1");

  % Add scopes
  add_block('simulink/Sinks/Scope', mdl + "/Scope", 'Position',[270 120 310 160]);
  add_line(mdl, "Mux/1", "Scope/1");

  set_param(mdl, 'StopTime', '40');
  save_system(mdl);
end

File: Tutorials/Codes/System_Dynamics_Course/Chapter20/Lesson4/Chapter20_Lesson4.py
# Chapter20_Lesson4.py
"""
Chapter 20 — Chaos, Complex Dynamics, and Computational Tools
Lesson 4 — Computational Tools: MATLAB/Simulink, Python (SciPy), and Modelica for System Dynamics

This script demonstrates:
1) Continuous-time ODE simulation (Lorenz system) with SciPy solve_ivp (adaptive RK and stiff solvers)
2) Event detection for Poincaré sections
3) Largest Lyapunov exponent estimation (Benettin-style) for the Lorenz flow (intro level)
4) Discrete-time map simulation (logistic map), bifurcation data, and Lyapunov exponent of the map
5) Reproducible export to CSV for post-processing / plotting in any tool

Dependencies:
- numpy
- scipy
"""

from __future__ import annotations
import math
import csv
from dataclasses import dataclass
from typing import Callable, Tuple

import numpy as np
from scipy.integrate import solve_ivp


# ----------------------------
# 1) Lorenz system (ODE)
# ----------------------------
@dataclass
class LorenzParams:
    sigma: float = 10.0
    rho: float = 28.0
    beta: float = 8.0 / 3.0


def lorenz_rhs(t: float, x: np.ndarray, p: LorenzParams) -> np.ndarray:
    """dx/dt = f(x)."""
    X, Y, Z = x
    return np.array([
        p.sigma * (Y - X),
        X * (p.rho - Z) - Y,
        X * Y - p.beta * Z
    ], dtype=float)


def lorenz_jacobian(x: np.ndarray, p: LorenzParams) -> np.ndarray:
    """Jacobian J = df/dx."""
    X, Y, Z = x
    return np.array([
        [-p.sigma, p.sigma, 0.0],
        [p.rho - Z, -1.0, -X],
        [Y, X, -p.beta]
    ], dtype=float)


def event_poincare_x0(t: float, x: np.ndarray, p: LorenzParams) -> float:
    """Event: x crosses 0 with positive direction."""
    return x[0]


event_poincare_x0.terminal = False
event_poincare_x0.direction = 1.0


def simulate_lorenz(
    x0: np.ndarray,
    t_span: Tuple[float, float],
    p: LorenzParams,
    method: str = "RK45",
    rtol: float = 1e-9,
    atol: float = 1e-12,
    max_step: float = 0.05,
) -> solve_ivp:
    """Adaptive integration with optional event handling."""
    fun = lambda t, x: lorenz_rhs(t, x, p)
    ev = lambda t, x: event_poincare_x0(t, x, p)
    sol = solve_ivp(
        fun=fun,
        t_span=t_span,
        y0=x0,
        method=method,
        rtol=rtol,
        atol=atol,
        max_step=max_step,
        events=ev,
        dense_output=True
    )
    if not sol.success:
        raise RuntimeError(f"Integration failed: {sol.message}")
    return sol


def export_trajectory_csv(path: str, t: np.ndarray, y: np.ndarray) -> None:
    """Write trajectory to CSV: t,x,y,z."""
    with open(path, "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(["t", "x", "y", "z"])
        for i in range(t.size):
            w.writerow([float(t[i]), float(y[0, i]), float(y[1, i]), float(y[2, i])])


def export_points_csv(path: str, pts: np.ndarray, header=("x", "y", "z")) -> None:
    with open(path, "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(list(header))
        for row in pts:
            w.writerow([float(v) for v in row])


# ----------------------------
# 2) Largest Lyapunov exponent (Lorenz, intro level)
#    Benettin-style: integrate tangent dynamics and renormalize.
# ----------------------------
def lce_lorenz_benettin(
    x0: np.ndarray,
    p: LorenzParams,
    t_transient: float = 5.0,
    t_total: float = 50.0,
    dt_renorm: float = 0.05,
    method: str = "RK45",
) -> float:
    """
    Estimate largest Lyapunov exponent λ_max.

    We integrate the extended system:
      x' = f(x)
      v' = J(x) v
    and periodically renormalize v to avoid overflow.

    Notes:
    - This is an educational implementation; for research-grade estimation,
      use multiple tangent vectors and QR orthonormalization.
    """
    # Transient integration to approach attractor
    sol_tr = simulate_lorenz(x0, (0.0, t_transient), p, method=method, max_step=dt_renorm)
    x = sol_tr.y[:, -1].astype(float)

    # Initial tangent vector
    v = np.array([1.0, 0.0, 0.0], dtype=float)
    v /= np.linalg.norm(v)

    # Integrate in chunks of dt_renorm
    t = 0.0
    sum_log = 0.0
    n = 0

    def ext_rhs(t_local: float, z: np.ndarray) -> np.ndarray:
        x_local = z[0:3]
        v_local = z[3:6]
        dx = lorenz_rhs(t_local, x_local, p)
        J = lorenz_jacobian(x_local, p)
        dv = J @ v_local
        return np.concatenate([dx, dv])

    while t < t_total:
        z0 = np.concatenate([x, v])
        sol = solve_ivp(
            fun=ext_rhs,
            t_span=(0.0, dt_renorm),
            y0=z0,
            method=method,
            rtol=1e-9,
            atol=1e-12,
            max_step=dt_renorm
        )
        if not sol.success:
            raise RuntimeError(f"Extended integration failed: {sol.message}")
        z1 = sol.y[:, -1]
        x = z1[0:3]
        v = z1[3:6]
        norm_v = float(np.linalg.norm(v))
        if norm_v == 0.0:
            # extremely unlikely; re-seed
            v = np.array([1.0, 0.0, 0.0], dtype=float)
            norm_v = 1.0
        sum_log += math.log(norm_v)
        n += 1
        v = v / norm_v
        t += dt_renorm

    return sum_log / (n * dt_renorm)


# ----------------------------
# 3) Logistic map (discrete-time)
# ----------------------------
def logistic_map(r: float, x0: float, n: int) -> np.ndarray:
    """x_{k+1} = r x_k (1 - x_k), returns array length n+1."""
    xs = np.empty(n + 1, dtype=float)
    x = float(x0)
    xs[0] = x
    for k in range(n):
        x = r * x * (1.0 - x)
        xs[k + 1] = x
    return xs


def logistic_bifurcation_data(
    r_values: np.ndarray,
    x0: float = 0.2,
    n_transient: int = 1000,
    n_keep: int = 200,
) -> np.ndarray:
    """
    Return (r, x) pairs after discarding transient.
    Useful for bifurcation diagrams.
    """
    rows = []
    for r in r_values:
        xs = logistic_map(float(r), x0, n_transient + n_keep)
        tail = xs[-n_keep:]
        for x in tail:
            rows.append((float(r), float(x)))
    return np.array(rows, dtype=float)


def logistic_lyapunov(r: float, x0: float = 0.2, n_transient: int = 1000, n: int = 5000) -> float:
    """
    Largest Lyapunov exponent for logistic map:
      λ = lim (1/N) Σ ln | f'(x_k) |, with f'(x)=r(1-2x)
    """
    xs = logistic_map(r, x0, n_transient + n)
    xs = xs[n_transient:]  # drop transient
    deriv = np.abs(r * (1.0 - 2.0 * xs))
    # Avoid log(0)
    deriv = np.maximum(deriv, 1e-300)
    return float(np.mean(np.log(deriv)))


# ----------------------------
# 4) Main demo
# ----------------------------
def main() -> None:
    out_dir = "outputs_ch20_l4"
    os.makedirs(out_dir, exist_ok=True)

    p = LorenzParams(sigma=10.0, rho=28.0, beta=8.0/3.0)
    x0 = np.array([1.0, 1.0, 1.0], dtype=float)

    # Lorenz simulation with adaptive RK (Dormand–Prince-like)
    sol = simulate_lorenz(
        x0=x0,
        t_span=(0.0, 40.0),
        p=p,
        method="RK45",
        max_step=0.02
    )
    export_trajectory_csv(os.path.join(out_dir, "lorenz_traj.csv"), sol.t, sol.y)

    # Poincaré section points (events at x=0 crossing)
    if sol.t_events and len(sol.t_events[0]) > 0:
        t_ev = sol.t_events[0]
        y_ev = sol.sol(t_ev) if sol.sol is not None else None
        if y_ev is not None:
            pts = y_ev.T  # rows
            export_points_csv(os.path.join(out_dir, "lorenz_poincare_x0.csv"), pts, header=("x", "y", "z"))

    # Largest Lyapunov exponent (educational)
    lce = lce_lorenz_benettin(x0=x0, p=p, t_transient=5.0, t_total=60.0, dt_renorm=0.05)
    with open(os.path.join(out_dir, "lorenz_lce.txt"), "w") as f:
        f.write(f"lambda_max ~ {lce:.6f}\n")

    # Logistic map: bifurcation data and Lyapunov
    r_vals = np.linspace(2.5, 4.0, 400)
    bif = logistic_bifurcation_data(r_vals, x0=0.2, n_transient=800, n_keep=120)
    np.savetxt(os.path.join(out_dir, "logistic_bifurcation.csv"), bif, delimiter=",", header="r,x", comments="")

    lyap_rows = []
    for r in np.linspace(2.8, 4.0, 200):
        lyap_rows.append((float(r), logistic_lyapunov(float(r), x0=0.2, n_transient=1000, n=4000)))
    np.savetxt(os.path.join(out_dir, "logistic_lyapunov.csv"), np.array(lyap_rows), delimiter=",", header="r,lambda", comments="")

    print("Done. Outputs written to:", out_dir)
    print("Lorenz LCE estimate:", lce)


if __name__ == "__main__":
    import os
    main()
